### PR TITLE
journald: support ordering journal entries by boot ID if no RTC is available

### LINKED ISSUE #662: sd-journal: find a smart way to order log lines from different boots if neither has a useful wallclock time.
I use a RaspberryPi2 with Archlinux and systemd version 219 to 222 (since the beginning of June until now).
The RaspberryPi does not have an RTC (real time clock).
The missing RTC results in the first log messages of every boot to be of the same date and more or less from the same time of day. Here you see an excerpt from my logs as seen with `journalctl`.
(The ellipsis were inserted by myself.)

```
-- Reboot --
Jan 01 01:00:07 gemipi systemd-journal[117]: Runtime journal is using 4.5M (max allowed 36.6M, trying to leave 54.9M free of 361.8M available → current limit 36.6M).
Jan 01 01:00:07 gemipi systemd-journal[117]: Permanent journal is using 104.0M (max allowed 256.0M, trying to leave 2.1G free of 11.7G available → current limit 256.0M).
Jan 01 01:00:07 gemipi systemd-journal[117]: Time spent on flushing to /var is 5.756ms for 2 entries.
...
-- Reboot --
Jan 01 01:00:08 gemipi systemd-journal[117]: Runtime journal is using 4.5M (max allowed 36.6M, trying to leave 55.0M free of 362.3M available → current limit 36.6M).
Jan 01 01:00:08 gemipi systemd-journal[117]: Permanent journal is using 40.0M (max allowed 1.4G, trying to leave 2.1G free of 12.2G available → current limit 1.4G).
Jan 01 01:00:08 gemipi systemd-journal[117]: Time spent on flushing to /var is 5.777ms for 2 entries.
...
-- Reboot --
Jan 01 01:00:08 gemipi systemd-journal[126]: Runtime journal is using 4.5M (max allowed 36.6M, trying to leave 54.9M free of 361.8M available → current limit 36.6M).
-- Reboot --
Jan 01 01:00:08 gemipi kernel: raid6: neonx2     394 MB/s
...
-- Reboot --
Jan 01 01:00:08 gemipi systemd-journal[126]: Permanent journal is using 88.0M (max allowed 256.0M, trying to leave 2.1G free of 11.7G available → current limit 256.0M).
-- Reboot --
Jan 01 01:00:08 gemipi systemd[1]: Mounted POSIX Message Queue File System.
-- Reboot --
Jan 01 01:00:08 gemipi systemd-journal[126]: Time spent on flushing to /var is 7.557ms for 2 entries.
...
-- Reboot --
Jan 01 01:00:08 gemipi systemd[1]: Mounted Debug File System.
-- Reboot --
Jan 01 01:00:08 gemipi kernel: raid6: using algorithm neonx4 (433 MB/s)
Jan 01 01:00:08 gemipi kernel: raid6: using intx1 recovery algorithm
-- Reboot --
Jan 01 01:00:08 gemipi systemd[1]: Mounted Temporary Directory.
-- Reboot --
Jan 01 01:00:08 gemipi kernel: bcm2708-dmaengine 3f007000.dma: DMA legacy API manager at f3007000, dmachans=0x7f35
```

As you can see, there are many boots which have only one line of logs, which do not even start with the usual "Runtime journal is ..." line.
This is certainly not correct.

Furthermore, at the first boot permanent journal is using 104.0M, at the second boot permanent journal is using only 40.0M. That does not make sense, chronologically. The messages were probably generated in the reverse chronological order.
I think that journald should guarantee that log-messages which are generated in some chronological order are displayed by journalctl in the same chronological order.

I suspect that `journalctl` is mixing up different boots when merging different `.journal` files.

### PR DESCRIPTION
**Pull request by poettering**
This fixes #662. Previously, on systems that lack a battery-buffered RTC we'd start up with basically a clock of zero, and then interleave entries from subsequent boots with this.

With this change we'll instead try to order boots by their boot ID, deriving order information form the timestamps of the newest entry we are aware of of that boot ID.

This should make things just work, even for old journal files, as long as a "roughly monotonic" clock is achieved at least once during each boot. This is what timesyncd should provide either based on a successful NTP fix or via restoring time from a timestamp file on disk.

This is on top of #26265 and #26213. Only the 6 topmost patches are actually new in this PR.

Fixes: #662

### CODE REVIEW

**Comment by yuwata** [comment]
needs-rebase

**Comment by bluca** [comment]
Looks related:

```
[  596.717464] systemd[1]: systemd-journald.service: Main process exited, code=killed, status=9/KILL
[  596.717528] systemd[1]: systemd-journald.service: Failed with result 'signal'.
```

File: test/TEST-04-JOURNAL/test.sh [hunk 1]
```diff
@@ -7,4 +7,11 @@ TEST_DESCRIPTION="Journal-related tests"
 # shellcheck source=test/test-functions
 . "${TEST_BASE_DIR:?}/test-functions"
 
+test_append_files() {
+    local workspace="${1:?}"
+
+    mkdir -p "$workspace/test-journals/"
+    cp -av "${TEST_BASE_DIR:?}/test-journals/"* "$workspace/test-journals/"
```
Comment (bluca) [inline test/TEST-04-JOURNAL/test.sh:14]: any chance you can make these test files (way) smaller, or better generate them on the fly? Currently it's an extra 56MB added to the git repo, which is quite a bit

Comment (poettering) [reply test/TEST-04-JOURNAL/test.sh:14]: i'd really like this to be real-life examples.

anyway, will compress them, that makes this 1M

Comment (bluca) [reply test/TEST-04-JOURNAL/test.sh:14]: Can you do zstd instead of gz? Total goes from 2.1M to 1.8M, or 1.6M with `zstd -19`

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
Sounds related on CentOS:

```
11:17:14 FAILED TASKS:
11:17:14 -------------
11:17:14 TEST-04-JOURNAL
```

**Comment by mrc0mmand** [comment]
> Sounds related on CentOS:
> 
> ```
> 11:17:14 FAILED TASKS:
> 11:17:14 -------------
> 11:17:14 TEST-04-JOURNAL
> ```

This looks like C8S is missing zstd. Lemme patch it up and restart the job.

**Comment by mrc0mmand** [comment]
> > > Sounds related on CentOS:
> > > ```
> > > 11:17:14 FAILED TASKS:
> > > 11:17:14 -------------
> > > 11:17:14 TEST-04-JOURNAL
> > > ```
> > 
> > 
> > This looks like C8S is missing zstd. Lemme patch it up and restart the job.
> 
> Note that doing that will automatically start compressing all journals/cores with zstd, in case you missed it there's a new compile option to let you select the compressor `-Ddefault-compression` in case you want to keep the rest of the CI infra using the current format

CentOS CI has been using zstd as well, the issue here is that the zstd utilities are in a different package - zstd (whereas the zstd libs are in libzstd(-devel)).

**Comment by bluca** [comment]
ah I didn't notice the test was manually decompressing - shouldn't journalctl be able to handle the compressed files just fine?

**Pull request merged**